File: SecondYear/semestrul4/MPP/PROIECT/CSharpProjectNetworking/Server/service/MainService.cs
using System.Collections.ObjectModel;
using System.Collections.Concurrent;
using Services;
using Model;

namespace Server;

public class MainService : IProjectServices
{
    private readonly IService<int, Organizator> organizatorService;
    private readonly IService<int, Participant> participantService;
    private readonly IService<int, Proba> probaService;
    private readonly IService<Tuple<Participant, Proba>, Inscriere> inscriereService;
    private readonly IService<int, CategorieVarsta> categorieVarstaService;

    private readonly ObservableCollection<ProbaDTO> probeStatisticsData;

    private Organizator currentOrganizator;

    private readonly IDictionary<string, IProjectObserver> loggedClients;
    // private readonly ConcurrentDictionary<string, IProjectObserver> _loggedInClients;
    private readonly object _lockObject = new object();
    public MainService(
        IService<int, Organizator> organizatorService,
        IService<int, Participant> participantService,
        IService<int, Proba> probaService,
        IService<Tuple<Participant, Proba>, Inscriere> inscriereService,
        IService<int, CategorieVarsta> categorieVarstaService)
    {
        this.organizatorService = organizatorService;
        this.participantService = participantService;
        this.probaService = probaService;
        this.inscriereService = inscriereService;
        this.categorieVarstaService = categorieVarstaService;
        
        probeStatisticsData = new ObservableCollection<ProbaDTO>();
        this.loggedClients = new Dictionary<string, IProjectObserver>();
        // _loggedInClients = new ConcurrentDictionary<string, IProjectObserver>();
        Console.WriteLine("MainService initialized."); 
        RefreshProbeStatistics(); 
    }

    public Organizator Login(string username, string password, IProjectObserver client)
    {
        lock (this)
        {
            Console.WriteLine($"Users now logged: {loggedClients.Count}");
            Organizator? org = Autentificator.Login(username, password);
            Console.WriteLine($"Authenticating user: {org.Username}, password: {org.Password}");

            if (org != null)
            {
                
                if (loggedClients.ContainsKey(username))
                {
                    Console.WriteLine($"User {username} is already logged in.");
                    throw new ProjectException("User already logged in.");
                }

                if (org.Password == password)
                {
                    loggedClients.Add(username, client);
                    Console.WriteLine($"User {username} logged in successfully. Observer added."); 
                    
                    return org;
                }
                else
                {
                    Console.WriteLine($"Failed to add user {username} to logged in clients (concurrent issue?)."); 
                    throw new ProjectException("Login failed due to an internal error.");
                }
            }
            else
            {
                Console.WriteLine($"User not found or authentication failed for: {username}"); 
                throw new ProjectException("Authentication failed! User not found or incorrect password.");
            }
        }
        
        
    }
    public void Logout(Organizator user, IProjectObserver client)
    {
        lock (this)
        {
            IProjectObserver localClient = loggedClients[user.Username];
            Console.WriteLine($"Attempting logout for user: {user.Username}"); 
            if (localClient == null)
            {
                Console.WriteLine($"User {user.Username} is not logged in."); // Logging substitut
            }
            
            loggedClients.Remove(user.Username);
        }
    }

    private void NotifyClients(Inscriere inscriere)
    {
        Console.WriteLine($"Preparing to notify clients about inscriere: {inscriere?.IDParticipant?.Nume} -> {inscriere?.IDProba?.Tip}"); // Logging substitut
        
        foreach (IProjectObserver client in loggedClients.Values)
        {
            
            Task.Run(() =>
            {
                try
                {
                    client.InscriereAdded(inscriere); 
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error during notification for client {client.GetType().Name}: {ex.Message}"); // Logging substitut
                }
            });
        }
        Console.WriteLine($"Submitted notification tasks for {inscriere?.IDParticipant?.Nume} to {loggedClients.Count} clients."); // Logging substitut
    }

    public IEnumerable<Organizator> GetAllOrganizatori()
    {
        lock (_lockObject)
        {
            try
            {
                return organizatorService.FindAll();
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to find all organizatori: {e.Message}"); // Logging substitut
                throw new ProjectException("Failed to find all organizatori.", e);
            }
        }
        
    }

    public IEnumerable<Proba> GetAllProba()
    {
        lock (_lockObject)
        {
            try
            {
                return probaService.FindAll();
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to find all probe: {e.Message}"); // Logging substitut
                throw new ProjectException("Failed to find all probe.", e);
            }
        }
        
    }

    public IList<ProbaDTO> GetAllProbaDto()
    {
        lock (_lockObject)
        {
            return new List<ProbaDTO>(probeStatisticsData);
        }
    }


    public void RefreshProbeStatistics()
    {
            Console.WriteLine("Refreshing probe statistics...");
            try
            {
                var newData = new List<ProbaDTO>();
                IEnumerable<Proba> probeList = GetAllProba();

                foreach (Proba proba in probeList)
                {
                    int count = CountParticipantsForProba(proba);
                    newData.Add(new ProbaDTO(proba, count));
                }

                probeStatisticsData.Clear();
                foreach (ProbaDTO item in newData)
                {
                    probeStatisticsData.Add(item);
                }

                Console.WriteLine($"Probe statistics refreshed. {probeStatisticsData.Count} items.");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to refresh probe statistics: {e.Message}\n{e.StackTrace}");
            }
        
    }


    private int CountParticipantsForProba(Proba proba)
    {
        if (proba == null) 
            return 0;

        var inscrieri = inscriereService.FindAll() ?? new List<Inscriere>(); 
        int count = 0;

        foreach (var inscriere in inscrieri)
        {
            if (inscriere?.IDProba?.Id != null && inscriere.IDProba.Id.Equals(proba.Id))
            {
                count++;
            }
        }

        return count;
    }


    public Proba GetProbaByName(string name)
    {
        lock (_lockObject)
        {
            try
            {
                if (probaService is ProbaService)
                {
                    return ((ProbaService) probaService).FindProbaByName(name); 
                }
                Console.WriteLine($"ProbaService instance not available for GetProbaByName. Service type: {probaService.GetType().Name}"); // Logging
                return null; 
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to find proba by name '{name}': {e.Message}"); // Logging substitut
                throw new ProjectException($"Failed to find proba by name: {name}", e);
            }
        }
        
    }

    public Proba GetProbaByNameAndRange(string name, int range)
    {
        lock (_lockObject)
        {
            try
            {
                if (probaService is ProbaService ps)
                {
                    return ps.FindProbaByNameAndRange(name, range); 
                }
                Console.WriteLine($"ProbaService instance not available for GetProbaByNameAndRange. Service type: {probaService.GetType().Name}"); // Logging
                return null; 
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to find proba by name '{name}' and rangeId '{range}': {e.Message}"); // Logging substitut
                throw new ProjectException($"Failed to find proba by name: {name} and rangeId: {range}", e);
            } 
        }
        
    }

    public CategorieVarsta GetVarstaByRange(string selectedCategory)
    {
        lock (_lockObject)
        {
            try
            {
                if (categorieVarstaService is CategorieVarstaService cvs)
                {
                    return cvs.FindVarstaByRange(selectedCategory);
                }
                Console.WriteLine($"CategorieVarstaService instance not available. Service type: {categorieVarstaService.GetType().Name}"); // Logging
                return null;
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to find varsta by category '{selectedCategory}': {e.Message}"); // Logging substitut
                throw new ProjectException($"Failed to find varsta by category: {selectedCategory}", e);
            } 
        }
        
    }

    public IList<ParticipantDTO> SearchParticipants(Proba selectedProba, CategorieVarsta selectedCategory)
    {
        lock (_lockObject)
        {
            try
            {
                var results = new List<ParticipantDTO>();
                var inscrieri = inscriereService.FindAll(); // Poate arunca excepție

                foreach (var inscriere in inscrieri)
                {
                    Proba? proba = inscriere.IDProba;
                    Participant? participant = inscriere.IDParticipant;

                    if (proba != null && participant != null &&
                        proba.Tip != null && selectedProba.Tip != null && proba.Varsta != null && // Null checks
                        proba.Tip.Equals(selectedProba.Tip) && proba.Varsta.Id.Equals(selectedCategory.Id))
                    {
                        if (participant.Cnp != null)
                        {
                            int age = CalculateAgeFromCNP(participant.Cnp);
                            results.Add(new ParticipantDTO(participant.Nume ?? "N/A", age)); // Folosim ?? pentru Nume null
                        }
                    }
                }
                return results;
            }
            catch (Exception e)
            {
                Console.WriteLine($"Failed to search participants for category '{selectedCategory.Id}': {e.Message}"); // Logging substitut
                throw new ProjectException($"Failed to search participants for category: {selectedCategory.Id}", e);
            }
        }
        
    }

    private int CalculateAgeFromCNP(string cnp)
    {
        // CNP format: SAALLZZJJNNNC
        int year;
        int firstDigit = int.Parse(cnp[0].ToString());
        int yearDigits = int.Parse(cnp.Substring(1, 2));

        if (firstDigit == 1 || firstDigit == 2)
        {
            year = 1900 + yearDigits;
        }
        else if (firstDigit == 5 || firstDigit == 6)
        {
            year = 2000 + yearDigits;
        }
        else
        {
            year = 1900 + yearDigits;
        }

        int currentYear = DateTime.Now.Year;
        return currentYear - year;
    }

    public string RegisterParticipant(string name, string cnp, string event1, string event2, string range)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return "Name must not be null or empty";
        }

        if (string.IsNullOrWhiteSpace(cnp) || cnp.Length != 13)
        {
            return "CNP must have 13 characters";
        }

        if (event1.Equals(event2))
        {
            return "You need to select different events!";
        }

        int age = CalculateAgeFromCNP(cnp);

        CategorieVarsta varsta1 = GetVarstaByRange(range);
        Proba proba1 = GetProbaByNameAndRange(event1, varsta1.Id);

        if (IsAgeInCategory(age, varsta1))
        {
            return "Not in the age range!";
        }

        Proba proba2 = GetProbaByNameAndRange(event2, varsta1.Id);

        lock (_lockObject)
        {
            Participant participant = FindParticipantByCNP(cnp);
            if (participant != null)
            {
                int existingRegistrations = CountRegistrationsForParticipant(participant);
                if (existingRegistrations >= 2 || (existingRegistrations == 1 && event2 != " "))
                {
                    return "Already registered at two events!";
                }
            }
            else
            {
                participant = new Participant(name, cnp);
                participantService.Save(participant);
            }

            var inscriere1 = new Inscriere(participant, proba1);
            inscriereService.Save(inscriere1);
            RefreshProbeStatistics();
            NotifyClients(inscriere1);

            if (event2 != " ")
            {
                var inscriere2 = new Inscriere(participant, proba2);
                inscriereService.Save(inscriere2);
                RefreshProbeStatistics();
                NotifyClients(inscriere2);
                
            }

            
        }
        
        return "Registration successful!";
    }

    public Participant FindParticipantByCNP(string cnp)
    {
        lock (_lockObject)
        {
            var participants = participantService.FindAll();
            foreach (var p in participants)
            {
                if (p.Cnp.Equals(cnp))
                {
                    return p;
                }
            }
            return null;
        }
        
    }

    private int CountRegistrationsForParticipant(Participant participant)
    {
        lock (_lockObject)
        {
            var inscrieri = inscriereService.FindAll();
            int count = 0;
            foreach (var inscriere in inscrieri)
            {
                if (inscriere.IDParticipant.Id.Equals(participant.Id))
                {
                    count++;
                }
            }
            return count;
        }
        
    }

    private bool IsAgeInCategory(int age, CategorieVarsta category)
    {
        return age < category.VarstaMin || age > category.VarstaMax;
    }

}

File: SecondYear/semestrul4/MPP/PROIECT/JavaProjectORM/ProjectServer/src/main/java/ro/mpp2025/javaprojectui/service/MainService.java
package ro.mpp2025.javaprojectui.service;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2025.javaprojectui.*;
import ro.mpp2025.javaprojectui.dto.ParticipantDTO;
import ro.mpp2025.javaprojectui.dto.ProbaDTO;
import ro.mpp2025.javaprojectui.Participant;

import java.sql.SQLException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainService implements ProjectServices {

    private final Service<Integer, Organizator> organizatorService;
    private final Service<Integer, Participant> participantService;
    private final Service<Integer, Proba> probaService;
    private final Service<Tuplu<Participant, Proba>, Inscriere> inscriereService;
    private final Service<Integer, CategorieVarsta> categorieVarstaService;

    private final ObservableList<ProbaDTO> probeStatisticsData = FXCollections.observableArrayList();

    private final Map<String, ProjectObserver> loggedInClients;
    private  ExecutorService notificationExecutor;

    private Organizator currentOrganizator;

    private static final Logger logger = LogManager.getLogger(MainService.class);

    public MainService(Service<Integer, Organizator> organizatorService, Service<Integer, Participant> participantService, Service<Integer, Proba> probaService, Service<Tuplu<Participant, Proba>, Inscriere> inscriereService, Service<Integer, CategorieVarsta> categorieVarstaService) {
        this.organizatorService = organizatorService;
        this.participantService = participantService;
        this.probaService = probaService;
        this.inscriereService = inscriereService;
        this.categorieVarstaService = categorieVarstaService;

        this.loggedInClients = new ConcurrentHashMap<>();
//        this.notificationExecutor = Executors.newFixedThreadPool(5);
    }

    @Override
    public synchronized Organizator login(String username, String password, ProjectObserver client) throws Exception {
        logger.info("Attempting login for user: {}", username);

        Optional<Organizator> orgOpt = Optional.ofNullable(Authenticator.login(username, password));
        if (orgOpt.isPresent()) {
            Organizator org = orgOpt.get();
            if (org.getPassword().equals(password)) {
                if (loggedInClients.containsKey(username)) {
                    logger.warn("User {} is already logged in.", username);
                    throw new ProjectException("User already logged in.");
                }
                loggedInClients.put(username, client);

                logger.info("User {} logged in successfully. Observer added.", username);
                return org;
            } else {
                logger.warn("Incorrect password for user: {}", username);
                throw new ProjectException("Authentication failed! Incorrect password.");
            }
        } else {
            logger.warn("User not found: {}", username);
            throw new ProjectException("Authentication failed! User not found.");
        }
    }

    @Override
    public synchronized void logout(Organizator user, ProjectObserver client) throws ProjectException {
        logger.info("Attempting logout for user: {}", user.getUsername());
        ProjectObserver removedClient = loggedInClients.remove(user.getUsername());
        if (removedClient == null) {
            logger.warn("User {} was not logged in or already logged out.", user.getUsername());
            throw new ProjectException("User " + user.getUsername() + " is not logged in.");
        }
        logger.info("User {} logged out successfully. Observer removed.", user.getUsername());
    }


    private void notifyClients(Inscriere inscriere) {
//        //////
        notificationExecutor = Executors.newFixedThreadPool(5);
        for (ProjectObserver client : loggedInClients.values()) {
            notificationExecutor.execute(() -> {
                try {
                    client.inscriereAdded(inscriere);
                    logger.info("Successfully notified client (observer) about update: {}", client.getClass().getName());
                } catch (Exception e) {
                    logger.error("Unexpected error during notification for client (observer) {}: {}", client.getClass().getName(), e.getMessage(), e);
                }
            });
        }
        notificationExecutor.shutdown();

    }

    @Override
    public Iterable<Organizator> getAllOrganizatori() throws ProjectException {
        try{
            logger.debug("Getting all organizatori...");
            var x =  organizatorService.findAll();
            logger.debug("Organizator found: {}", x);
            return x;
        }catch (Exception e){
            throw new ProjectException("Failed to find all organizatori." + e);
        }

    }

    @Override
    public Iterable<Proba> getAllProba() throws ProjectException {
        try{
            return probaService.findAll();
        }
        catch (Exception e){
            throw new ProjectException("Failed to find all probe." + e.getMessage());
        }
    }

    public Optional<Proba> getProbaById(Integer id) throws SQLException {
        logger.info("Getting proba by id: {}", id);
        return probaService.findOne(id);
    }

    public int createProba(Proba proba) throws SQLException {
        logger.info("Creating new proba: {}", proba);
        probaService.save(proba);
        return proba.getId();
    }

    public Optional<Proba> updateProba(Proba proba) throws SQLException {
        logger.info("Updating proba: {}", proba);
        return probaService.update(proba);
    }

    public void deleteProba(Integer probaId) throws SQLException {
        logger.info("Deleting proba: {}", probaId);
        probaService.delete(probaId);
    }

    @Override
    public synchronized List<ProbaDTO> getAllProbaDTO() throws ProjectException {
        try{
            if (probeStatisticsData == null) {
                return new ArrayList<>();
            }
            return new ArrayList<>(probeStatisticsData);
        }catch (Exception e){
            throw new ProjectException("Failed to find all probedto:" + e.getMessage());
        }

    }

    public synchronized void refreshProbeStatistics() throws ProjectException {
        try{

            List<ProbaDTO> newData = new ArrayList<>();
            Iterable<Proba> probe = getAllProba();

            for (Proba proba : probe) {
                int count = countParticipantsForProba(proba);
                newData.add(new ProbaDTO(proba, count));
            }

            probeStatisticsData.setAll(newData);
        }catch (Exception e){
            throw new ProjectException("Failed to refreshData:" + e.getMessage());
        }

    }

    public int countParticipantsForProba(Proba proba) {
        Iterable<Inscriere> inscrieri = inscriereService.findAll();
        int count = 0;
        for (Inscriere inscriere : inscrieri) {
            if (inscriere.getIDproba().getId().equals(proba.getId())) {
                count++;
            }
        }
        return count;
    }

    @Override
    public Proba getProbaByName(String name) throws ProjectException {
        try{
            if (probaService instanceof ProbaService) {
                return ((ProbaService) probaService).findProbaByName(name);
            }
            return null;
        }catch (Exception e){
            throw new ProjectException("Failed to find probaby by name: " + name);
        }

    }

    @Override
    public Proba getProbaByNameAndRange(String name, Integer range) throws ProjectException {
        try {
            if (probaService instanceof ProbaService) {
                return ((ProbaService) probaService).findProbaByNameAndRange(name, range);
            }
            return null;
        }catch (Exception e){
            throw new ProjectException("Failed to find probaby by name: " + name + " range: " + range);
        }

    }


    @Override
    public CategorieVarsta getVarstaByRange(String selectedCategory) throws ProjectException {
        try{
            if (categorieVarstaService instanceof CategorieVarstaService) {
                return ((CategorieVarstaService) categorieVarstaService).findVarstaByRange(selectedCategory);
            }
            return null;
        }catch (Exception e){
            throw new ProjectException("Failed to find varsta by selectedCategory: " + selectedCategory);
        }

    }


    @Override
    public List<ParticipantDTO> searchParticipants(Proba selectedProba, CategorieVarsta selectedCategory) throws ProjectException {
        try{
            List<ParticipantDTO> results = new ArrayList<>();
            Iterable<Inscriere> inscrieri = inscriereService.findAll();

            for (Inscriere inscriere : inscrieri) {
                Proba proba = inscriere.getIDproba();
                Participant participant = inscriere.getIDparticipant();

                if(proba.getTip().equals(selectedProba.getTip()) && proba.getVarsta().getId().equals(selectedCategory.getId())){
                    int age = calculateAgeFromCNP(participant.getCnp());
                    results.add(new ParticipantDTO(participant.getNume(), age));
                }
            }

            return results;
        }catch (Exception e){
            throw new ProjectException("Failed to find participants by selectedCategory: " + selectedCategory);
        }

    }

    private int calculateAgeFromCNP(String cnp) {

        int year;
        int firstDigit = Character.getNumericValue(cnp.charAt(0));
        int yearDigits = Integer.parseInt(cnp.substring(1, 3));

        if (firstDigit == 1 || firstDigit == 2) {
            year = 1900 + yearDigits;
        } else if (firstDigit == 5 || firstDigit == 6) {
            year = 2000 + yearDigits;
        } else {
            year = 1900 + yearDigits;
        }

        int currentYear = LocalDate.now().getYear();
        return currentYear - year;
    }

    @Override
    public synchronized String registerParticipant(String name, String cnp, String event1, String event2, String range) throws ProjectException, SQLException {

        if (name == null || name.trim().isEmpty()) {
            return "Name must not be null or empty";
        }

        if (cnp == null || cnp.trim().isEmpty() || cnp.length() != 13) {
            return "CNP must have 13 characters";
        }

        if (event1 == null) {
            return "You need to select at least one event!";
        }

        if(event1.equals(event2)){
            return "You need to select different events!";
        }

        int age = calculateAgeFromCNP(cnp);


        CategorieVarsta varsta1 = getVarstaByRange(range);
        Proba proba1 = getProbaByNameAndRange(event1, varsta1.getId());

        System.out.println(proba1);
        System.out.println(varsta1);

        if (isAgeInCategory(age, varsta1)) {
            return "Not in the age range!";
        }

        Proba proba2 = getProbaByNameAndRange(event2, varsta1.getId());

        System.out.println(proba2);


        Optional<Participant> existingParticipant = findParticipantByCNP(cnp);
        Participant participant;

        if (existingParticipant.isPresent()) {
            participant = existingParticipant.get();
            int existingRegistrations = countRegistrationsForParticipant(participant);
            if (existingRegistrations >= 2 || (existingRegistrations == 1 && event2 != null)) {
                return "Already registered at two events!";
            }
        } else {
            participant = new Participant(name, cnp);
            participantService.save(participant);

        }

        synchronized (this) {


            Inscriere inscriere1 = new Inscriere(participant, proba1);
            inscriereService.save(inscriere1);

            refreshProbeStatistics();

            logger.info("Notifying {} clients about the update for Proba ID {}", loggedInClients.size(), proba1.getId());
            notifyClients(inscriere1);

            if (event2 != null) {
                Inscriere inscriere2 = new Inscriere(participant, proba2);
                inscriereService.save(inscriere2);

                refreshProbeStatistics();

                logger.info("Notifying {} clients about the update for Proba ID {}", loggedInClients.size(), proba2.getId());
                notifyClients(inscriere2);
            }
        }


        return "Registration successful!";
    }

    @Override
    public Optional<Participant> findParticipantByCNP(String cnp) throws ProjectException{
        try{
            Iterable<Participant> participants = participantService.findAll();
            for (Participant p : participants) {
                if (p.getCnp().equals(cnp)) {
                    return Optional.of(p);
                }
            }
            return Optional.empty();
        } catch (Exception e) {
            throw new ProjectException(e.getMessage());
        }

    }

    private synchronized int countRegistrationsForParticipant(Participant participant) {
        Iterable<Inscriere> inscrieri = inscriereService.findAll();
        int count = 0;
        for (Inscriere inscriere : inscrieri) {
            if (inscriere.getIDparticipant().getId().equals(participant.getId())) {
                count++;
            }
        }
        return count;
    }

    private boolean isAgeInCategory(int age, CategorieVarsta category) {
        return age < category.getVarstaMin() || age > category.getVarstaMax();
    }


}
